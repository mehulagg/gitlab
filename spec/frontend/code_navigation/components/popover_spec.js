import { shallowMount } from '@vue/test-utils';
import Popover from '~/code_navigation/components/popover.vue';
import DocLine from '~/code_navigation/components/doc_line.vue';

const DEFINITION_PATH_PREFIX = 'http://gitlab.com';

const MOCK_CODE_DATA = Object.freeze({
  hover: [
    {
      language: 'javascript',
      tokens: [
        [
          {
            class: 'k',
            value: 'function',
          },
          {
            value: ' main() {',
          },
        ],
        [
          {
            value: '}',
          },
        ],
      ],
    },
  ],
  definition_path: 'test.js',
  definitionLineNumber: 20,
});

const MOCK_DOCS_DATA = Object.freeze({
  hover: [
    {
      language: null,
      value: 'console.log',
    },
  ],
  definition_path: 'test.js#L20',
});

let wrapper;

function factory({ position, data, definitionPathPrefix, blobPath = 'index.js' }) {
  wrapper = shallowMount(Popover, {
    propsData: { position, data, definitionPathPrefix, blobPath },
    stubs: { DocLine },
  });
}

describe('Code navigation popover component', () => {
  afterEach(() => {
    wrapper.destroy();
  });

  it('renders popover', () => {
    factory({
      position: { x: 0, y: 0, height: 0 },
      data: MOCK_CODE_DATA,
      definitionPathPrefix: DEFINITION_PATH_PREFIX,
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders link with hash to current file', () => {
    factory({
      position: { x: 0, y: 0, height: 0 },
      data: MOCK_CODE_DATA,
      definitionPathPrefix: DEFINITION_PATH_PREFIX,
      blobPath: 'test.js',
    });

    expect(wrapper.find('[data-testid="go-to-definition-btn"]').attributes('href')).toBe('#L20');
  });

  describe('code output', () => {
    it('renders code output', () => {
      factory({
        position: { x: 0, y: 0, height: 0 },
        data: MOCK_CODE_DATA,
        definitionPathPrefix: DEFINITION_PATH_PREFIX,
      });

      expect(wrapper.find({ ref: 'code-output' }).exists()).toBe(true);
      expect(wrapper.find({ ref: 'doc-output' }).exists()).toBe(false);
    });
  });

  describe('documentation output', () => {
    it('renders code output', () => {
      factory({
        position: { x: 0, y: 0, height: 0 },
        data: MOCK_DOCS_DATA,
        definitionPathPrefix: DEFINITION_PATH_PREFIX,
      });

      expect(wrapper.find({ ref: 'code-output' }).exists()).toBe(false);
      expect(wrapper.find({ ref: 'doc-output' }).exists()).toBe(true);
    });
  });
});
