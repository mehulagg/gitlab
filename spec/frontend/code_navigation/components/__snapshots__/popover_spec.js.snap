// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code navigation popover component renders popover 1`] = `
<div
  class="popover code-navigation-popover popover-font-size-normal gl-popover bs-popover-bottom show"
  style="left: 0px; top: 0px;"
>
  <div
    class="arrow"
    style="left: 0px;"
  />
   
  <div
    class="overflow-auto code-navigation-popover-container"
  >
    <div
      class=""
    >
      <pre
        class="border-0 bg-transparent m-0 code highlight text-wrap"
      >
        <span
          class="line"
          lang="javascript"
        >
          <span
            class="k"
          >
            function
          </span>
          <span>
             main() {
          </span>
        </span>
        <span
          class="line"
          lang="javascript"
        >
          <span>
            }
          </span>
        </span>
      </pre>
    </div>
  </div>
   
  <div
    class="popover-body border-top"
  >
    <gl-button-stub
      category="tertiary"
      class="w-100"
      data-testid="go-to-definition-btn"
      href="http://gitlab.com/test.js"
      icon=""
      size="medium"
      target="_blank"
      variant="default"
    >
      
      Go to definition
    
    </gl-button-stub>
  </div>
</div>
`;
